haystack/agents/base.py

```diff
@@ -0,0 +1,298 @@
+            if final_answer is not None:
+                return self._format_answer(
+                    query=query,
+                    answer=final_answer,
+                    tool_names=tool_names,
+                    tool_names_with_descriptions=tool_names_with_descriptions,
+                    generated_text=transcript + pred,
+                )
+            tool_name, tool_input = self._extract_tool_name_and_tool_input(pred=pred)
+            if tool_name is None or tool_input is None:
+                raise AgentError("Wrong output format.")
+
+            result = self._run_tool(tool_name=tool_name, tool_input=tool_input, params=params)
+            observation = self._extract_observation(result)
+            transcript += f"{pred}\nObservation: {observation}\nThought: Now that I know that {observation} is the answer to {tool_name} {tool_input}, I "
+
+        logger.warning(
+            "Maximum number of agent iterations (%s) reached for query (%s). Increase max_iterations "
+            "or no answer can be provided for this query.",
+            max_iterations,
+            query,
+        )
+        return self._format_answer(
+            query=query,
+            answer="",
+            tool_names=tool_names,
+            tool_names_with_descriptions=tool_names_with_descriptions,
+            generated_text=transcript,
+        )
+
+    def run_batch(
+        self, queries: List[str], max_iterations: Optional[int] = None, params: Optional[dict] = None
+    ) -> Dict[str, str]:
+        """
+        Runs the Agent in a batch mode
+
+        :param queries: List of search queries.
+        :param max_iterations: The number of times the Agent can run a tool plus then once infer it knows the final answer.
+            Set at least to 2 so that the Agent can run one tool and then infer it knows the final answer. Default 5.
+        :param params: A dictionary of parameters that you want to pass to those tools that are pipelines.
+                       To pass a parameter to all nodes in those pipelines, use: `{"top_k": 10}`.
+                       To pass a parameter to targeted nodes in those pipelines, use:
+                        `{"Retriever": {"top_k": 10}, "Reader": {"top_k": 3}}`.
+                        Parameters can only be passed to tools that are pipelines but not nodes.
+        """
+        results: Dict = {"queries": [], "answers": [], "transcripts": []}
+        for query in queries:
+            result = self.run(query=query, max_iterations=max_iterations, params=params)
+            results["queries"].append(result["query"])
+            results["answers"].append(result["answers"])
+            results["transcripts"].append(result["transcript"])
+
+        return results
+
+    def _run_tool(
+        self, tool_name: str, tool_input: str, params: Optional[dict] = None
+    ) -> Union[Tuple[Dict[str, Any], str], Dict[str, Any]]:
+        if not self.has_tool(tool_name):
+            raise AgentError(
```

Comment:
Should we include transcripts in AgentError by default? It could be useful I think

Reply:
Would be useful I agree. 👍 I will check how to do that. One downside is that if the Agent generates text with `$` in it, filling the template will probably fail. And filling the template is required to have a proper transcript.